src/Corax/Utils/EntryIdEncodings.cs

```diff
@@ -0,0 +1,138 @@
+    }
+
+    /// <summary>
+    /// Container id (page) without encodings.
+    /// </summary>
+    public static long GetContainerId(long entryId)
+    {
+        return (entryId >> EntryIdOffset);
+    }
+
+    [MethodImpl(MethodImplOptions.AggressiveInlining | MethodImplOptions.AggressiveOptimization)]
+    public static unsafe void Encode(Span<long> entries, Span<short> frequencies, bool usePtrVersion = false)
+    {
+        Debug.Assert(frequencies.Length == entries.Length);
+
+        if (usePtrVersion == false)
+            goto Classic;
+
+        for (int idX = 0; idX < entries.Length; ++idX)
+        {
+            ref var entryId = ref Unsafe.Add(ref MemoryMarshal.GetReference(entries), idX);
+            ref var frequency = ref Unsafe.Add(ref MemoryMarshal.GetReference(frequencies), idX);
+            entryId = entryId << EntryIdOffset | FrequencyQuantization(frequency) << ContainerTypeOffset;
+        }
+
+        return;
+
+        Classic:
+        for (int i = 0; i < entries.Length; ++i)
+        {
+            var ecd = FrequencyQuantization(frequencies[i]) << ContainerTypeOffset | entries[i] << EntryIdOffset;
+            entries[i] = ecd;
+        }
+    }
+
+    [MethodImpl(MethodImplOptions.AggressiveInlining | MethodImplOptions.AggressiveOptimization)]
+    public static long DecodeAndDiscardFrequency(long entryId) => entryId >> EntryIdOffset;
+
+    [MethodImpl(MethodImplOptions.AggressiveInlining | MethodImplOptions.AggressiveOptimization)]
+    public static void DecodeAndDiscardFrequency(Span<long> entries, int read)
+    {
+        for (int i = 0; i < read; ++i)
+        {
+            entries[i] >>= EntryIdOffset;
+        }
+    }
+
+    public static void Decode(Span<long> matches, Span<short> frequencies)
+    {
+        for (int i = 0; i < matches.Length; ++i)
+        {
+            frequencies[i] = FrequencyReconstructionFromQuantization((short)((matches[i] >> ContainerTypeOffset) & Mask));
+            matches[i] >>= EntryIdOffset;
+        }
+    }
+
+    // 4 bit - exp (as 2^(4 + 2i) per step
+    // 4 bit - mantissa
+    private static readonly long[] Step = Enumerable.Range(0, 16).Select(i => (long)Math.Pow(2, 4 + 2 * i)).ToArray();
+
+    private static readonly long[] StepSize = Enumerable.Range(0, 16).Select(i =>
+        i == 0
+            ? 16
+            : (Step[i] - Step[i - 1])).ToArray();
+
+    private static readonly double[] LevelSizeInStep = StepSize.Select(i => i / 16D).ToArray();
+
+    private static readonly short[] FrequencyTable = Enumerable.Range(0, byte.MaxValue).Select(i => FrequencyReconstructionFromQuantizationFromFunction(i)).ToArray();
+    
+    internal static long FrequencyQuantization(short frequency)
+    {
+        if (frequency < 16) //shortcut
+            return frequency;
+
+        int level = 0;
+        for (level = 0; level < 16 && frequency >= Step[level]; ++level)
+        {
+        } // look up for range
```

Comment:
This can be done with a find the higher set bit and a lookup table too to avoid the `for` altogether and much packed assembler.

Reply:
Marked as todo